### PR TITLE
fix: use base64 as image provider

### PR DESCRIPTION
**Pull request by haritianaadriano**

### CODE REVIEW

File: src/main/java/school/hei/haapi/model/bean/Base64EncoderBean.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.stereotype.Component;
+
+@Component
+public class Base64EncoderBean {
+  @Bean
+  public static Base64.Encoder Base64Encoder() {
+    return Base64.getEncoder();
+  }
+}
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/model/bean/Base64EncoderBean.java]: remove this file as we are not sure whether Base64.Encoder is Thread-safe or not

File: src/main/java/school/hei/haapi/service/utils/Base64Converter.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+@Component
+public class Base64Converter implements Function<ClassPathResource, String> {
+
+  @Override
+  @SneakyThrows
+  public String apply(ClassPathResource path) {
+    try {
+      byte[] file = path.getInputStream().readAllBytes();
+      return Base64EncoderBean.Base64Encoder().encodeToString(file);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/Base64Converter.java]: use Base64.getEncoder() here

File: src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import org.springframework.core.io.Resource;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ClassPathResourceResolver implements BiFunction<String, String, ClassPathResource> {
+  @Override
+  public ClassPathResource apply(String fileName, String fileType) {
+    Resource resource = new ClassPathResource("static/images/" + fileName + fileType);
+    return (ClassPathResource) resource;
+  }
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java]: return Resource not ClassPathResource

File: src/main/resources/static/styles/main.css [hunk 4]
```diff
@@ -18,7 +18,7 @@ a {
   width: 500px;
   height: 150px;
   margin: 0 auto;
-  background-image: url(../assets/images/HEI_logo.png);
+  background-image: url(../images/HEI_logo.png);
```
Comment (Mahefaa) [inline src/main/resources/static/styles/main.css]: do you still need this ?

File: src/main/resources/static/styles/main.css [hunk 5]
```diff
@@ -76,7 +76,7 @@ blockquote {
   display: block;
   width: 150px;
   height: 150px;
-  background-image: url(../assets/images/signature.png);
+  background-image: url(../images/signature.png);
```
Comment (Mahefaa) [inline src/main/resources/static/styles/main.css]: same question

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java [hunk 6]
```diff
@@ -6,10 +6,10 @@
...
 
 @Component
-public class ClassPathResourceResolver implements BiFunction<String, String, ClassPathResource> {
+public class ClassPathResourceResolver implements BiFunction<String, String, Resource> {
   @Override
-  public ClassPathResource apply(String fileName, String fileType) {
+  public Resource apply(String fileName, String fileType) {
     Resource resource = new ClassPathResource("static/images/" + fileName + fileType);
-    return (ClassPathResource) resource;
+    return resource;
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java]: return new

File: src/main/java/school/hei/haapi/service/utils/Base64Converter.java [hunk 7]
```diff
@@ -3,22 +3,23 @@
...
 import school.hei.haapi.model.exception.ApiException;
 
 @Component
-public class Base64Converter implements Function<ClassPathResource, String> {
+public class Base64Converter implements Function<Resource, String> {
 
   @Override
   @SneakyThrows
-  public String apply(ClassPathResource path) {
+  public String apply(Resource path) {
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/Base64Converter.java]: not path !!!

File: src/main/java/school/hei/haapi/service/StudentFileService.java [hunk 8]
```diff
@@ -32,8 +33,8 @@ public byte[] generatePdf(String studentId, String template) {
 
   private Context loadContext(String studentId) {
     ClassPathResourceResolver pathResolver = new ClassPathResourceResolver();
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/StudentFileService.java]: why is it a method field ?

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/StudentFileService.java (hunk 8 above)
Comment (haritianaadriano) [reply src/main/java/school/hei/haapi/service/StudentFileService.java]: it's to access to apply method

File: src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java (hunk 6 above)
Comment (haritianaadriano) [reply src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java]: how we'll access to the file ? if we return a new Resource

File: src/main/java/school/hei/haapi/service/StudentFileService.java (hunk 8 above)
Comment (Mahefaa) [reply src/main/java/school/hei/haapi/service/StudentFileService.java]: use it as an instance atrribute in the StudentFileService Class

File: src/main/java/school/hei/haapi/service/StudentFileService.java [hunk 9]
```diff
@@ -28,16 +32,24 @@ public byte[] generatePdf(String studentId, String template) {
   }
 
   private Context loadContext(String studentId) {
+    ClassPathResourceResolver pathResolver = new ClassPathResourceResolver();
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/StudentFileService.java]: Also a suitable variable name would be "resourceResolver" or "classPathResourceResolver'
pathResolver induces that you only resolve the path and that you get a path

File: src/main/java/school/hei/haapi/service/StudentFileService.java [hunk 10]
```diff
@@ -28,16 +32,24 @@ public byte[] generatePdf(String studentId, String template) {
...
+    Resource logo = pathResolver.apply("HEI_logo", ".png");
+    Resource signature = pathResolver.apply("signature", ".png");
+    Base64Converter base64Converter = new Base64Converter();
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/StudentFileService.java]: put base64Converter as Class instance field

File: src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.util.function.BiFunction;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.Resource;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ClassPathResourceResolver implements BiFunction<String, String, Resource> {
+  @Override
+  public Resource apply(String fileName, String fileType) {
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/ClassPathResourceResolver.java]: filename and extension as parameters

**Review by Mahefaa — CHANGES_REQUESTED** [review]
few changes left, good job :))

File: src/test/java/school/hei/haapi/integration/StudentFileIT.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                .build(),
+            HttpResponse.BodyHandlers.ofByteArray());
+
+    assertNotNull(response);
+  }
+
+  @Test
+  void manager_load_certificate_ok() throws IOException, InterruptedException {
+    String STUDENT_CERTIFICATE = "/students/"+STUDENT2_ID+"/scholarship_certificate/raw";
+    HttpClient httpClient = HttpClient.newBuilder().build();
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/StudentFileIT.java:88]: use newClient()

File: src/test/java/school/hei/haapi/integration/StudentFileIT.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                .GET()
+                .header("Authorization", "Bearer " + MANAGER1_TOKEN)
+                .build(),
+            HttpResponse.BodyHandlers.ofByteArray());
+
+    assertNotNull(response);
+  }
+
+  private static ApiClient anApiClient(String token) {
+    return TestUtils.anApiClient(token, StudentIT.ContextInitializer.SERVER_PORT);
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/StudentFileIT.java]: change to StudentFileIT

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/utils/Base64Converter.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.Resource;
+import org.springframework.stereotype.Component;
+import school.hei.haapi.model.exception.ApiException;
+
+@Component
+public class Base64Converter implements Function<Resource, String> {
+
+  @Override
+  @SneakyThrows
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/utils/Base64Converter.java]: rm this

**Review by Mahefaa — CHANGES_REQUESTED** [review]
format your code

File: src/main/java/school/hei/haapi/service/StudentFileService.java [hunk 15]
```diff
@@ -28,16 +33,22 @@ public byte[] generatePdf(String studentId, String template) {
   }
 
   private Context loadContext(String studentId) {
+    Resource logo = classPathResourceResolver.apply("HEI_logo", ".png");
+    Resource signature = classPathResourceResolver.apply("signature", ".png");
     Context context = new Context();
     User student =
         userRepository
             .findById(studentId)
             .orElseThrow(() -> new NotFoundException("Student not found"));
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/StudentFileService.java]: use the corresponding method from UserService

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/StudentFileService.java [hunk 16]
```diff
@@ -28,16 +33,19 @@ public byte[] generatePdf(String studentId, String template) {
...
 
   private Context loadContext(String studentId) {
+    Resource logo = classPathResourceResolver.apply("HEI_logo", ".png");
+    Resource signature = classPathResourceResolver.apply("signature", ".png");
     Context context = new Context();
-    User student =
-        userRepository
-            .findById(studentId)
-            .orElseThrow(() -> new NotFoundException("Student not found"));
+    User student = userService.getById(studentId);
```
Comment (Mahefaa) [inline src/main/java/school/hei/haapi/service/StudentFileService.java]: interchange these lines

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/test/java/school/hei/haapi/integration/StudentFileIT.java [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+        httpClient.send(
+            HttpRequest.newBuilder()
+                .uri(URI.create(basePath + STUDENT_CERTIFICATE))
+                .GET()
+                .header("Authorization", "Bearer " + STUDENT1_TOKEN)
+                .build(),
+            HttpResponse.BodyHandlers.ofByteArray());
+
+    assertEquals(HttpStatus.OK.value(), response.statusCode());
+    assertNotNull(response);
```
Comment (Mahefaa) [inline src/test/java/school/hei/haapi/integration/StudentFileIT.java:82]: test the responseBody also,
test if its length is > 0

**Review by Mahefaa — CHANGES_REQUESTED** [review]

**Pull request merged**